### ISSUE TITLE
Qualitative Data Categorisation Improvements

### ISSUE THREAD

**Issue by benwallace2nhs** [nhsengland/datascience#361]
### Project Title

Qualitative Data Categorisation Improvements

### One-Line Summary

LLM-based multi-label text classifier for NHS patient experience comments, replacing a legacy ML model with a RAG-augmented approach requiring no model training

### Origin (Optional)

Insight and Voice

### Project Tags

- [ ] Domain: Primary care
- [ ] Domain: Secondary care
- [ ] Domain: Social Care
- [ ] Domain: Emergency care
- [ ] Domain: Diseases
- [ ] Domain: Public/Population Health
- [ ] Domain: Workforce & Allocation
- [ ] Domain: Prescribing
- [ ] Domain: Financial
- [ ] Technique: Forecasting
- [x] Technique: Classification
- [ ] Technique: Computer Vision
- [ ] Technique: Deep Learning
- [x] Technique: Large Language Models (LLM)
- [x] Technique: Machine Learning
- [ ] Technique: Neural Networks
- [x] Technique: Natural Language Processing (NLP)
- [ ] Technique: Linkage
- [ ] Technique: Simulation
- [ ] Type: Best Practice Guidance
- [ ] Type: Data Validation
- [ ] Type: Documentation
- [ ] Type: Ethics
- [ ] Type: Explainability
- [ ] Type: Generation
- [ ] Type: Linkage
- [x] Type: Modelling
- [ ] Type: Reproducible Analytical Pipeline (RAP)
- [ ] Type: Research
- [ ] Data Type: Structured
- [x] Data Type: Unstructured
- [ ] Data Type: Multi-modal
- [ ] Data Type: Sound/Audio
- [ ] Data Type: Genomics
- [ ] Data Type: Video/Image
- [x] Data Type: Text Data
- [ ] Data Type: Time Series
- [ ] Data Access: Patient Identifiable
- [ ] Data Access: Synthetic
- [ ] Data Access: Open Data
- [ ] Data Access: Pseudonymised
- [x] Language: Python
- [ ] Language: SQL
- [ ] Language: R
- [ ] Language: Web Development (e.g. Javascript, HTML, CSS)
- [ ] Status: Work in Progress
- [ ] Status: Complete
- [ ] Status: On Pause
- [x] Status: In Development
- [ ] Status: Retired
- [ ] Status: Deployed
- [ ] Status: Experimental

### Project Description

<img width="2085" height="1185" alt="Image" src="https://github.com/user-attachments/assets/ee4e9491-3f3b-43f9-b25c-c46d5c22c6d2" />

Patient experience teams across the NHS collect thousands of free-text comments via the Friends and Family Test (FFT). These comments need categorising against 31 themes from the Qualitative Data Categorisation (QDC) Framework to identify patterns and drive service improvement.

This is a multi-label classification problem — a single comment can be assigned several categories simultaneously. For example, "The staff were great but I couldn't get parked" covers both "Staff manner" and "Parking".

The legacy approach used a trained sklearn/BERT ensemble that required manual retraining and achieved ~0.70 weighted F1.

We replaced this with an LLM-based classifier using Claude Sonnet via the NHS Federated Data Platform (FDP) API. The system uses retrieval-augmented few-shot prompting — for each comment being classified, it retrieves semantically similar examples from a corpus of 13,000 labelled comments and includes them as in-context calibration. Comments are processed in batches of 50 per API call for efficiency. This approach requires no model training, and making category changes is as simple as updating a prompt.

Beyond classification, the LLM approach enables segmented sentiment highlighting — each comment is broken into its component clauses, with each segment assigned its own category and sentiment. This means "The staff were great but I couldn't get parked" produces two segments: "The staff were great" (Staff manner, Positive) and "I couldn't get parked" (Parking, Negative). This gives patient experience teams a much richer, more actionable view of feedback than flat category labels alone.

## Results

The LLM approach achieves:

- **Macro F1: 0.776** (vs ~0.70 for the legacy model)
- **Weighted F1: 0.777**
- Multi-label classification across 31 categories with no training data pipeline
- Per-segment sentiment and evidence extraction

### Relevant Links (Optional)

Open Source Code & Documentation (Currently Private): https://github.com/nhsengland/QDC_LLM
Technical Methodology (Currently Private): https://github.com/nhsengland/QDC_LLM/blob/main/docs/methodology.md
Original Framwork Documentation: https://github.com/The-Strategy-Unit/PatientExperience-QDC

### Project Status

Current Project

### Main Work Area Category

Natural Language Processing Products
